homeassistant/util/logging.py

```diff
@@ -119,28 +119,28 @@ def log_exception(format_err: Callable[..., Any], *args: Any) -> None:
 @overload
 def catch_log_exception(
     func: Callable[..., Coroutine[Any, Any, Any]], format_err: Callable[..., Any]
-) -> Callable[..., Coroutine[Any, Any, None]]:
+) -> Callable[..., Coroutine[Any, Any, None]] | Callable[..., None]:
     """Overload for Callables that return a Coroutine."""
 
 
 @overload
 def catch_log_exception(
     func: Callable[..., Any], format_err: Callable[..., Any]
-) -> Callable[..., None | Coroutine[Any, Any, None]]:
+) -> Callable[..., Coroutine[Any, Any, None]] | Callable[..., None]:
     """Overload for Callables that return Any."""
 
 
 def catch_log_exception(
     func: Callable[..., Any], format_err: Callable[..., Any]
-) -> Callable[..., None | Coroutine[Any, Any, None]]:
+) -> Callable[..., Coroutine[Any, Any, None]] | Callable[..., None]:
```

Comment:
This seems weird. The doc string for this type is that it is to decorate a callback. Why would it then ever return a coroutine?

Reply:
> The doc string for this type is that it is to decorate a callback. Why would it then ever return a coroutine?

It doesn't. The return type is always a callable. However the callback can also be a coroutine-function in which case an async function should also be returned. I.e. `Callable[..., Coroutine[Any, Any, None]]`

https://github.com/home-assistant/core/blob/0ab19fe6f6349f1af9e9eea17c8275a6d9756d38/homeassistant/util/logging.py#L144-L155

https://github.com/home-assistant/core/blob/0ab19fe6f6349f1af9e9eea17c8275a6d9756d38/homeassistant/util/logging.py#L171

Reply:
I thought that we were using overloads here to clearly specify that coroutine in -> coroutine out. Callback in -> callback out. 

A callback cannot be a coroutine function, as they are 2 different types https://developers.home-assistant.io/docs/asyncio_categorizing_functions

Reply:
> I thought that we were using overloads here to clearly specify that coroutine in -> coroutine out. Callback in -> callback out.

Exactly. I got the callback name mixed up. In the context of this function, and I believe that's the intended meaning of the docstring, it should probably just refer to `func` which **can** be an async function.

The function itself is only used in two places. In both async functions can be passed as first argument.
https://github.com/home-assistant/core/blob/5501b7e710bed9ec32b2afc195a75f37eab1bf69/homeassistant/helpers/dispatcher.py#L65-L79

https://github.com/home-assistant/core/blob/5501b7e710bed9ec32b2afc195a75f37eab1bf69/homeassistant/components/mqtt/client.py#L209-L220